kyo-sttp/jvm/src/main/scala/kyo/interop/Adapter.scala

```diff
@@ -0,0 +1,401 @@
+package kyo.interop
+
+import java.util.concurrent.Flow.*
+import kyo.*
+import kyo.Emit.Ack
+import kyo.Maybe.*
+import kyo.Result.*
+import kyo.kernel.ArrowEffect
+import scala.annotation.tailrec
+
+object Adapters:
+
+    private trait KyoSubscriber[A] extends Subscriber[A]:
+        def isDone: Boolean < IO
+        def await: Unit < (Async & Abort[SubscriberDone])
+        def interrupt: Unit < IO
+    end KyoSubscriber
+
+    abstract class SubscriberDone
+    case object SubscriberDone extends SubscriberDone
+
+    private def makeSubscriber[A](
+        bufferSize: Int
+    )(
+        using Frame
+    ): (KyoSubscriber[A], Fiber.Promise[Nothing, (Subscription, Queue[A])]) < IO =
+        IO.Unsafe.apply {
+            val queue              = Queue.Unsafe.init[A](capacity = bufferSize, access = Access.MultiProducerMultiConsumer)
+            val onSubscribePromise = Fiber.Promise.Unsafe.init[Nothing, (Subscription, Queue[A])]()
+            val kyoSubscriber = new KyoSubscriber[A]:
+                val isSubscribedOrInterrupted = AtomicBoolean.Unsafe.init(false)
+
+                // volatile is enough, as these below can only be set by the producer thread
+                @volatile
+                private var _isDone: Maybe[Result[Throwable, Unit]] = Maybe.empty
+                @volatile
+                private var _waitPromise: Maybe[Fiber.Promise.Unsafe[SubscriberDone, Unit]] = Maybe.empty
+
+                override def isDone: Boolean < IO = IO(_isDone.isDefined)
+
+                override def interrupt: Unit < IO = isSubscribedOrInterrupted.safe.set(true)
+
+                private def failDone(doneResult: Result[Throwable, Unit]): Nothing < Abort[SubscriberDone] =
+                    doneResult.fold { error =>
+                        Abort.panic(error.getFailure)
+                    } { _ =>
+                        Abort.fail(SubscriberDone)
+                    }
+                end failDone
+
+                override def await: Unit < (Async & Abort[SubscriberDone]) =
+                    _isDone.fold {
+                        val awaitPromise = Fiber.Promise.Unsafe.init[SubscriberDone, Unit]()
+                        _waitPromise = Maybe(awaitPromise)
+                        queue.empty().fold { _ =>
+                            // subscriber is not done, but queue is closed
+                            _waitPromise = Maybe.empty
+                            fail(new IllegalStateException("Queue is closed"))
+                            await
+                        } { isEmpty =>
+                            if isEmpty then
+                                _isDone.fold {
```

Comment:
Do you mind using pattern matching with `Absent`/`Present` instead of `fold`? It's the pattern across the codebase